x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SLMGetExpiredSnapshotsAction.java

```diff
@@ -0,0 +1,312 @@
+            super(INSTANCE.name(), actionFilters, transportService.getTaskManager());
+            this.repositoriesService = repositoriesService;
+            final var threadPool = transportService.getThreadPool();
+            this.retentionExecutor = threadPool.executor(ThreadPool.Names.MANAGEMENT);
+            this.threadContext = threadPool.getThreadContext();
+        }
+
+        private static class ResultsBuilder {
+            private final Map<String, List<Tuple<SnapshotId, String>>> resultsByRepository = ConcurrentCollections.newConcurrentMap();
+
+            Response getResponse() {
+                return new Response(Map.copyOf(resultsByRepository));
+            }
+
+            void addResult(String repository, List<Tuple<SnapshotId, String>> snapshotsToDelete) {
+                if (snapshotsToDelete.isEmpty()) {
+                    assert resultsByRepository.containsKey(repository) == false;
+                } else {
+                    final var previousValue = resultsByRepository.put(repository, snapshotsToDelete);
+                    assert previousValue == null : repository + ": " + previousValue + " vs " + snapshotsToDelete;
+                }
+            }
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+            final var resultsBuilder = new ResultsBuilder();
+            try (var refs = new RefCountingRunnable(() -> listener.onResponse(resultsBuilder.getResponse()))) {
+                for (final var repositoryName : request.repositories()) {
+
+                    final Repository repository;
+                    try {
+                        repository = repositoriesService.repository(repositoryName);
+                    } catch (RepositoryMissingException e) {
+                        logger.debug("[{}]: repository not found", repositoryName);
+                        continue;
+                    }
+
+                    if (repository.isReadOnly()) {
+                        logger.debug("[{}]: skipping readonly repository", repositoryName);
+                        continue;
+                    }
+
+                    retentionExecutor.execute(
+                        ActionRunnable.wrap(
+                            refs.acquireListener(),
+                            perRepositoryListener -> SubscribableListener
+
+                                // Get repository data
+                                .newForked(repository::getRepositoryData)
+
+                                // Collect snapshot details by policy, and get any missing details by reading SnapshotInfo
+                                .<SnapshotDetailsByPolicy>andThen(
+                                    retentionExecutor,
+                                    threadContext,
+                                    (l, repositoryData) -> getSnapshotDetailsByPolicy(
+                                        repositoriesService.repository(repositoryName),
```

Comment:
Why reload the repository here rather than passing in `repository` which was calculated above?

Reply:
Thanks, that was an oversight :) d96ec2de81bb287a01be277e46c3c7be16c8adc4